api/types/stacks/stacks.go

```diff
@@ -0,0 +1,91 @@
+package stacks // import "github.com/docker/docker/api/types/stacks
+
+import (
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/swarm"
+)
+
+// Stack is an object representing a server-side stack
+type Stack struct {
+	// ID is the ID of the stack, generated by the engine
+	ID string
+	swarm.Meta
+
+	// Spec is the StackSpec defining what this stack should look like
+	Spec StackSpec `json:",omitempty"`
+}
+
+// StackSpec defines the spec for a server-side stack.
+type StackSpec struct {
+	// StackSpecs include swarm-style annotations. The stack name cannot be
+	// changed once it has been created.
+	swarm.Annotations
+
+	// Services is a list of all service specs desired for this stack
+	Services []swarm.ServiceSpec `json:",omitempty"`
+
+	// Networks is a list of NetworkCreateRequest that define the networks
+	// required for this stack. NetworkCreateRequest includes the network name.
+	Networks []types.NetworkCreateRequest `json:",omitempty"`
+
+	// Secrets is a list of all Secrets required for this stack
+	Secrets []swarm.SecretSpec `json:",omitempty"`
+
+	// Configs is a list of all Configs required for this stack
+	Configs []swarm.ConfigSpec `json:",omitempty"`
+}
+
+// TODO(dperny): the following types should go in ../client.go, but are here
+// for ease of early review
+
+type StackCreateOptions struct {
+	// EncodedRegistryAuth is the encoded registry authorization credentials
+	// used for services in this stack. This is identical to the
+	// EncodedRegistryAuth field in ServiceCreateOptions and
+	// ServiceUpdateOptions
+	//
+	// This field follows the format of the X-Registry-Auth header.
+	EncodedRegistryAuth string
+
+	// QueryRegistry indicates whether service operations need to contact a
+	// registry. This is identical to the QueryRegistry field of
+	// ServiceCreateOptions and ServiceUpdateOptions
+	QueryRegistry bool
+}
+
+// StackCreateResponse contains the information returned to a client on the
+// creation of a stack.
+type StackCreateResponse struct {
+	// ID is the ID of the created stack. The resources defined in the stack
+	// are created asynchronously
+	ID string
+
+	// TODO(dperny): should we include a "Warnings" field?
```

Comment:
We may want too indeed :angel:

Reply:
What sort of state should result in a create success with warnings as opposed to an error?

Reply:
I can't imagine any, which is why I've opted not to include it. Not including it doesn't close the door on at some point including warnings in the future. Currently, as far as I can tell, the warning is only used to warn about digest pinning.